mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +50,277 @@
-            es_host (str): The Elasticsearch host server
+            app_path (str): The application path.
+            resource_loader (ResourceLoader): An object which can load resources for the resolver.
+            entity_type (str): The entity type associated with this entity resolver.
+            er_config (dict): A classifier config
+            es_host (str): The Elasticsearch host server.
+            es_client (Elasticsearch): The Elasticsearch client.
         """
-        self._app_namespace = get_app_namespace(app_path)
-        self._resource_loader = resource_loader
-        self._normalizer = resource_loader.query_factory.normalize
+        er_config = (
+            kwargs.pop("er_config", None) or
+            get_classifier_config("entity_resolution", app_path=app_path)
+        )
+        er_config = _correct_deprecated_er_config(er_config)
+        resolver_type = er_config["model_settings"]["resolver_type"]
+        cls._validate_resolver_type(resolver_type)
+        return ENTITY_RESOLVER_MODEL_MAPPINGS.get(resolver_type)(
+            app_path, resource_loader, entity_type, er_config, **kwargs
+        )
+
+
+class EntityResolverModelBase(ABC):
+    """
+    Base class for Entity Resolvers
+    """
+
+    def __init__(self, app_path, resource_loader, entity_type, er_config):
+        """Initializes an entity resolver"""
+        self.app_path = app_path
         self.type = entity_type
         self._is_system_entity = Entity.is_system_entity(self.type)
-        self._exact_match_mapping = None
-        self._er_config = get_classifier_config("entity_resolution", app_path=app_path)
-        self._es_host = es_host
-        self._es_config = {"client": es_client, "pid": os.getpid()}
-        self.ready = False
+        self.name = er_config["model_settings"]["resolver_type"]
+        self.entity_map = resource_loader.get_entity_map(self.type)
+        self.normalizer = resource_loader.query_factory.normalize
 
-        if self._is_system_entity:
-            canonical_entities = []
-        else:
-            canonical_entities = self._resource_loader.get_entity_map(self.type).get(
-                "entities", []
-            )
-        self._no_canonical_entity_map = len(canonical_entities) == 0
+        # default configs useful for obtaining model cache path
+        for key, value in self.default_er_config.get("model_settings", {}).items():
+            er_config["model_settings"][key] = value
+        self.er_config = er_config
 
-    @property
-    def _es_index_name(self):
-        return EntityResolver.ES_SYNONYM_INDEX_PREFIX + "_" + self.type
+        self.dirty = False  # bool, True if exists any unsaved generated data that can be saved
+        self.ready = False  # bool, True if the model is fit by calling .fit()
 
     @property
-    def _use_text_rel(self):
-        return self._er_config["model_type"] == "text_relevance"
+    def default_er_config(self):
```

Comment:
I thought the default setting is populated? If its empty, I dont see why we need a function to create an empty value dict

Reply:
It is empty only in the base class. It's just a fallback option in case that method is not available in the derived class.